### PR TITLE
[Basic] Ensure empty NullTerminatedStringRef have valid pointer

### PR DESCRIPTION
**Pull request by rintaro**
`NullterminatedStringRef` is meant to be safe for reading `data()` as a valid C-string.

rdar://108215490

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Review by bnbarham — APPROVED** [review]

File: lib/Sema/TypeCheckMacros.cpp [hunk 1]
```diff
@@ -687,20 +691,15 @@ Expr *swift::expandMacroExpr(
   }
 
   /// The discriminator used for the macro.
-  std::string cachedDiscriminator;
-  auto getDiscriminator = [&]() -> StringRef {
-    if (!cachedDiscriminator.empty())
-      return cachedDiscriminator;
-
+  LazyValue<std::string> discriminator([&]() -> std::string {
```
Comment (rintaro) [inline lib/Sema/TypeCheckMacros.cpp:694]: Don't we have similar "lazy" value thing in llvm or swiftBasic? 🤔

**Comment by rintaro** [comment]
https://github.com/apple/swift/pull/65263

**Pull request merged**